pkg/repository/v1/payloadstore.go

```diff
@@ -0,0 +1,450 @@
+			Location:            PayloadLocationExternal,
+			ExternalLocationKey: &key,
+		}
+
+		marshaledContent, err := content.Marshal()
+
+		if err != nil {
+			return fmt.Errorf("failed to marshal external payload content: %w", err)
+		}
+
+		ids = append(ids, opt.Id)
+		insertedAts = append(insertedAts, opt.InsertedAt)
+		payloadsToOffload = append(payloadsToOffload, marshaledContent)
+	}
+
+	// TODO: Update payloads in the db in a tx-safe way here?
+	return p.queries.OffloadPayloadsToExternalStore(ctx, tx, sqlcv1.OffloadPayloadsToExternalStoreParams{
+		Tenantid:    sqlchelpers.UUIDFromStr(tenantId),
+		Ids:         ids,
+		Insertedats: insertedAts,
+		Values:      payloadsToOffload,
+	})
+}
+
+func (p *payloadStoreRepositoryImpl) ProcessPayloadWAL(ctx context.Context, tenantId string) (bool, error) {
+	ctx, span := telemetry.NewSpan(ctx, "process-payload-wal")
+	defer span.End()
+
+	tx, commit, rollback, err := sqlchelpers.PrepareTx(ctx, p.pool, p.l, 5000)
+
+	defer rollback()
+
+	if err != nil {
+		return false, fmt.Errorf("failed to prepare transaction: %w", err)
+	}
+
+	walRecords, err := p.queries.PollPayloadWALForRecordsToOffload(ctx, tx, sqlcv1.PollPayloadWALForRecordsToOffloadParams{
+		Tenantid:  sqlchelpers.UUIDFromStr(tenantId),
+		Polllimit: 1000,
+	})
+
+	if len(walRecords) == 0 {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	retrieveOpts := make([]RetrievePayloadOpts, len(walRecords))
+	offloadAts := make([]pgtype.Timestamptz, len(walRecords))
+	payloadIds := make([]int64, len(walRecords))
+	payloadInsertedAts := make([]pgtype.Timestamptz, len(walRecords))
+	payloadTypes := make([]string, len(walRecords))
+
+	for i, record := range walRecords {
+		retrieveOpts[i] = RetrievePayloadOpts{
+			Id:         record.PayloadID,
+			InsertedAt: record.PayloadInsertedAt,
+			Type:       record.PayloadType,
+		}
+
+		offloadAts[i] = record.OffloadAt
+		payloadIds[i] = record.PayloadID
+		payloadInsertedAts[i] = record.PayloadInsertedAt
+		payloadTypes[i] = string(record.PayloadType)
+	}
+
+	payloads, err := p.bulkRetrieve(ctx, tx, tenantId, retrieveOpts...)
+
+	if err != nil {
+		return false, err
+	}
+
+	externalStoreOpts := make([]StorePayloadOpts, 0)
+
+	for opts, payload := range payloads {
+		externalStoreOpts = append(externalStoreOpts, StorePayloadOpts{
+			Id:         opts.Id,
+			InsertedAt: opts.InsertedAt,
+			Type:       opts.Type,
+			Payload:    payload,
+		})
+	}
+
+	err = p.offloadToExternal(ctx, tx, tenantId, externalStoreOpts...)
+
+	if err != nil {
+		return false, err
+	}
+
+	err = p.queries.EvictPayloadWALRecords(ctx, tx, sqlcv1.EvictPayloadWALRecordsParams{
+		Tenantid:           sqlchelpers.UUIDFromStr(tenantId),
+		Payloadids:         payloadIds,
+		Payloadinsertedats: payloadInsertedAts,
+		Offloadats:         offloadAts,
+		Payloadtypes:       payloadTypes,
+	})
+
+	if err != nil {
+		return false, err
+	}
+
+	if err := commit(ctx); err != nil {
+		return false, err
+	}
+
+	return true, nil
```

Comment:
We're always returning `true` here which seems incorrect: this is signaling the operation to continue looping so this is effectively creating an infinite loop